### PR TITLE
[AutoDiff] Ban local-scope derivative registration.

### PR DESCRIPTION
**Pull request by rxwei**
Fix a long-time hole in derivative registration: local scopes with possible captures.  Derivative registration was not intended to support local scopes where a derivative function and an original function can have different closure contexts.  Improper use of derivative registration in local scopes can result in crashers deeper down the compiler, e.g. TF-881.

This PR bans derivative registration in local scopes by emitting diagnostics at parsing time. This applies to three attributes:
1. `@differentiable` with JVP or VJP specified. The `jvp:` and `vjp:` parameters in `@differentiable` will be banned as well as part of the retroactive derivative registration effort (TF-1001).
2. `@derivative(of:)`.
3. `@transpose(of:)` (today called `@transposing`).

The diagnostics are emitted from the parser because there are precedents: access modifiers in local scopes are banned this way.

Resolves SR-11849. Defines away TF-881.

### CODE REVIEW

**Comment by rxwei** [comment]
Note: Local-scope derivative registration is likely going to be discussed and supported in the distant future, after the first evolution proposal goes through.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: test/AutoDiff/transposing_attr_parse.swift [hunk 1]
```diff
@@ -64,3 +64,10 @@ func tfoo(t: Float) -> Float {
 func tfoo(t: Float) -> Float {
   return t
 }
+
+func localTransposeRegistration() {
+  // Not okay. Derivative registration can only be non-local.
```
Comment (dan-zheng) [inline test/AutoDiff/transposing_attr_parse.swift]:
```suggestion
  // Not okay. Transpose registration can only be non-local.
```

File: test/AutoDiff/compiler_crashers/tf881-derivative-local-variable-capture.swift [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: not --crash %target-swift-emit-silgen %s
+// RUN: %target-swift-emit-silgen %s -verify
```
Comment (dan-zheng) [inline test/AutoDiff/compiler_crashers/tf881-derivative-local-variable-capture.swift]: Please move this test from `test/AutoDiff/compiler_crashers` to `test/AutoDiff/compiler_crashers_fixed`.

**Review by dan-zheng — APPROVED** [review]
I wonder if there are issues when `@differentiable` is declared on a local-scope original function that captures local values without registering derivatives?

Answer: it seems to work fine:

```swift
func wrapper() {
  var free1: Int = 1
  var free2: Float = 2

  @differentiable
  func original(_ x: Float) -> Float {
    _ = free1
    return x * free2
  }
  print(gradient(at: 3, in: original))
}
wrapper()
// 2.0

// Original function:
// sil private [ossa] @original : $@convention(thin) (Float, Int, Float) -> Float

// Generated:
// sil hidden [ossa] @original__vjp_src_0_wrt_0 : $@convention(thin) (Float, Int, Float) -> (Float, @owned @callee_guaranteed (Float) -> Float)
// sil hidden [ossa] @original__pullback_src_0_wrt_0 : $@convention(thin) (Float, @owned $original_bb0__PB__src_0_wrt_0)
```

**Comment by rxwei** [comment]
> I wonder if there are issues when `@differentiable` is declared on a local-scope original function that captures local values without registering derivatives?

It's not an issue because the compiler-generated derivative/transpose will always have the same captures as the original function.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**